metadatastore/commands.py

```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
     beamline_id : str
-        Beamline String identifier. Not unique, just an indicator of
-        beamline code for multiple beamline systems
-    beamline_config : metadatastore.documents.Document or str
-        if Document:
-            The metadatastore beamline config document
-        if str:
-            uid of beamline config corresponding to a given run
-    uid : str, optional
-        Globally unique id string provided to metadatastore
+        Beamline String identifier.
+    uid : str
+        Globally unique id to identify this RunStart
     owner : str, optional
-        A username associated with the entry
+        A username associated with the RunStart
     group : str, optional
-        A group (e.g., UNIX group) associated with the entry
+        An experimental group associated with the RunStart
     project : str, optional
         Any project name to help users locate the data
     custom: dict, optional
         Any additional information that data acquisition code/user wants
-        to append to the Header at the start of the run.
+        to append to the RunStart at the start of the run.
 
     Returns
     -------
-    run_start: mongoengine.Document
-        Inserted mongoengine object
+    runstart : str
+        uid of the inserted document.  Use `runstart_given_uid` to get
+        the full document.
 
     """
-    if uid is None:
-        uid = str(uuid.uuid4())
     if custom is None:
         custom = {}
-    if owner is None:
-        owner = ''
-    if group is None:
-        group = ''
-    if project is None:
-        project = ''
-
-    beamline_config = _get_mongo_document(beamline_config, BeamlineConfig)
-    run_start = RunStart(time=time, scan_id=scan_id,
-                         time_as_datetime=_todatetime(time), uid=uid,
-                         beamline_id=beamline_id,
-                         beamline_config=beamline_config,
-                         owner=owner, group=group, project=project,
-                         **custom)
-
-    run_start.save(validate=True, write_concern={"w": 1})
-    logger.debug('Inserted RunStart with uid %s', run_start.uid)
+
+    runstart = RunStart(time=time, scan_id=scan_id,
+                        uid=uid,
+                        beamline_id=beamline_id,
+                        owner=owner, group=group, project=project,
+                        **custom)
+
+    runstart = runstart.save(validate=True, write_concern={"w": 1})
+
+    _cache_runstart(runstart.to_mongo().to_dict())
+    logger.debug('Inserted RunStart with uid %s', runstart.uid)
 
     return uid
 
 
 @_ensure_connection
-def insert_run_stop(run_start, time, uid=None, exit_status='success',
-                    reason=None, custom=None):
-    """ Provide an end to a sequence of events. Exit point for an
-    experiment's run.
+def insert_run_stop(run_start, time, uid, exit_status='success',
+                    reason='', custom=None):
+    """Insert RunStop document into database
```

Comment:
Is it worth noting that this is implemented so that only one `RunStop` can exist per `RunStart` and if a second one is created then it will raise?